core/trino-main/src/main/java/io/trino/util/DateTimeUtils.java

```diff
@@ -67,9 +69,70 @@ public static int parseDate(String value)
         // - the required format is 'YYYY-MM-DD'
         // - all components should be unsigned numbers
         // https://github.com/trinodb/trino/issues/10677
+        OptionalInt days = tryParseDateYyyyMmDd(value);
+        if (days.isPresent()) {
+            return days.getAsInt();
+        }
         return toIntExact(TimeUnit.MILLISECONDS.toDays(DATE_FORMATTER.parseMillis(value)));
     }
 
+    /**
+     * Check if the date is in format yyyy-MM-dd, if yes, return the epoch days; otherwise
+     * return OptionalInt.empty(), indicating the parse failed, and should fall back to generic
+     * parsing. Mixing the format checking and parsing in one pass for better performance.
+     * @param value date string to be parsed
+     * @return the number of days since 1970-01-01
+     */
+    public static OptionalInt tryParseDateYyyyMmDd(String value)
+    {
+        if (value.length() != 10) {
+            return OptionalInt.empty();
+        }
+        // year
+        int year = value.charAt(0) - '0';
+        if (year < 0 || year > 9) {
+            return OptionalInt.empty();
+        }
+        int curDigit = 0;
```

Comment:
you can move this variable inside the loop now

Reply:
done in 72e06703cbffd04afc38670c2f73c497db9c748b